Sources/Sentry/SentryNetworkTracker.m

```diff
@@ -0,0 +1,126 @@
+
+@interface
+SentryNetworkTracker ()
+
+@property (nonatomic, strong) SentryPerformanceTracker *tracker;
+
+@end
+
+@implementation SentryNetworkTracker
+
++ (SentryNetworkTracker *)sharedInstance
+{
+    static SentryNetworkTracker *instance = nil;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{ instance = [[self alloc] init]; });
+    return instance;
+}
+
+- (instancetype)init
+{
+    if (self = [super init]) {
+        self.tracker = SentryPerformanceTracker.shared;
+    }
+    return self;
+}
+
+- (void)urlSessionTaskResume:(NSURLSessionTask *)sessionTask
+{
+    NSURL *url = [[sessionTask currentRequest] URL];
+    if (url == nil)
+        return;
+
+    NSURL *apiUrl = [NSURL URLWithString:SentrySDK.options.dsn];
+    if ([url.host isEqualToString:apiUrl.host] && [url.path containsString:apiUrl.path])
+        return;
+
+    NSString *statePath = NSStringFromSelector(@selector(state));
+    [sessionTask addObserver:self
+                  forKeyPath:statePath
+                     options:NSKeyValueObservingOptionNew
+                     context:nil];
+
+    SentrySpanId *spanId = [self.tracker startSpanWithName:url.absoluteString
+                                                 operation:SENTRY_NETWORK_REQUEST_OPERATION];
+
+    objc_setAssociatedObject(sessionTask, &SENTRY_NETWORK_REQUEST_TRACKER_SPAN_ID, spanId,
+        OBJC_ASSOCIATION_RETAIN_NONATOMIC);
+}
+
+- (void)observeValueForKeyPath:(NSString *)keyPath
+                      ofObject:(id)object
+                        change:(NSDictionary<NSKeyValueChangeKey, id> *)change
+                       context:(void *)context
+{
+    if ([keyPath isEqualToString:NSStringFromSelector(@selector(state))]) {
+        NSURLSessionTask *sessionTask = object;
+        if (sessionTask.state != NSURLSessionTaskStateRunning) {
+            SentrySpanId *spanId
+                = objc_getAssociatedObject(sessionTask, &SENTRY_NETWORK_REQUEST_TRACKER_SPAN_ID);
+
+            if (spanId != nil) {
+                [self.tracker finishSpan:spanId withStatus:[self statusForSessionTask:sessionTask]];
+                [sessionTask removeObserver:self forKeyPath:NSStringFromSelector(@selector(state))];
+            }
+        }
+    }
+}
+
+- (SentrySpanStatus)statusForSessionTask:(NSURLSessionTask *)task
+{
+    switch (task.state) {
+    case NSURLSessionTaskStateSuspended:
+    case NSURLSessionTaskStateCanceling:
+        return kSentrySpanStatusCancelled;
+    case NSURLSessionTaskStateCompleted: {
+        if (task.response != nil && [task.response isKindOfClass:[NSHTTPURLResponse class]]) {
+            NSHTTPURLResponse *response = (NSHTTPURLResponse *)task.response;
+            return [self statusForResponse:response];
+        }
+    } break;
+    case NSURLSessionTaskStateRunning:
+        break;
+    }
+    return kSentrySpanStatusUndefined;
+}
+
+- (SentrySpanStatus)statusForResponse:(NSHTTPURLResponse *)response
```

Comment:
`l`: Maybe worth adding a link to `https://develop.sentry.dev/sdk/event-payloads/span/` as a comment?

Reply:
`m`: As we only use the `statusCode` of the request we could only pass the `statusCode` instead of the whole `repsonse` to this method. We could also rename it to something like
```suggestion
- (SentrySpanStatus)spanStatusForHttpResponseStatusCode:(NSInteger)statusCode
```